### PR TITLE
Boundary Equilibrium Loss (BEGAN)

### PR DESCRIPTION
**Pull request by martyn**
https://arxiv.org/pdf/1703.10717.pdf

BEGAN with support for WGAN or LSGAN(original research)

Preview:
image

### CODE REVIEW

File: examples/2d-measure-accuracy.py [hunk 1]
```diff
@@ -23,32 +23,50 @@ def parse_args():
...
     net = tf.concat(axis=0, values=[x,g])
+    original = net
     net = linear(net, 128, scope=prefix+'linone')
-    net = tf.nn.crelu(net)
-    net = linear(net, 128, scope=prefix+'linend')
+    net = tf.nn.relu(net)
+    net = linear(net, 2, scope=prefix+'linend')
+    
+    # works.  hyperparam? 
+    net = config.distance(original,net)
```
Comment (martyn) [inline examples/2d-measure-accuracy.py]: This only makes sense if you want to build an autoencoder.  I wonder how we can refactor this?

File: examples/static-batch-test.py [hunk 2]
```diff
@@ -69,9 +68,9 @@ def generator_config():
     )
 
 def discriminator_config():
-    return hg.discriminators.pyramid_discriminator.config(
+    return hg.discriminators.autoencoder_discriminator.config(
```
Comment (martyn) [inline examples/static-batch-test.py]: Some issue as 2d-trainer - we might want to search for something other than BEGAN.  Shouldn't hold up the release.

File: hypergan/config.py [hunk 3]
```diff
@@ -25,7 +25,7 @@ def selector(args):
     # Generator configuration
     selector.set("generator", [resize_conv_generator.config()])
 
-    selector.set("trainer", rmsprop_trainer.config())
+    selector.set("trainer", alternating_trainer.config())
```
Comment (martyn) [inline hypergan/config.py]: Config should default to BEGAN + LSGAN.

File: hypergan/discriminators/pyramid_discriminator.py [hunk 4]
```diff
@@ -108,30 +109,24 @@ def should_ignore_layer(layer, resize):
...
-      filter_size_w = 2
-      filter_size_h = 2
-      filter = [1,filter_size_w,filter_size_h,1]
-      stride = [1,filter_size_w,filter_size_h,1]
-      depth = int(int(net.get_shape()[3])*depth_increase)
-      if i ==0:
-          depth = config.first_conv_size
-      if strided:
-          net = conv2d(net, depth, name=prefix+'_expand_layer'+str(i), k_w=3, k_h=3, d_h=filter_size_h, d_w=filter_size_w, regularizer=None)
+      if config.foundation == 'additive':
```
Comment (martyn) [inline hypergan/discriminators/pyramid_discriminator.py]: Lets just remove this option and change it to always additive.  It's much cheaper and still gives good results.

File: hypergan/discriminators/pyramid_discriminator.py [hunk 5]
```diff
@@ -108,30 +109,24 @@ def should_ignore_layer(layer, resize):
...
 
-          net = conv2d(net, depth, name=prefix+'_expand_layer'+str(i), k_w=3, k_h=3, d_h=1, d_w=1, regularizer=None,gain=config.orthogonal_initializer_gain)
-
-          net = tf.nn.avg_pool(net, ksize=filter, strides=stride, padding='SAME')
+      if i ==0:
+          depth = config.first_conv_size
+      net = config.block(config, net, depth, prefix+'_layer_'+str(i)+'_')
 
-      print('[discriminator] layer', net)
+    print('[discriminator] layer', net)
```
Comment (martyn) [inline hypergan/discriminators/pyramid_discriminator.py]: I think this output is wrong.

File: hypergan/losses/common.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        with(tf.variable_scope("discriminator", reuse=reuse)):
+            ds = discriminator.create(gan, discriminator, net, net, gan.graph.xs, gan.graph.gs,prefix="d_"+str(i))
+            bs = gan.config.batch_size
+            net = ds
+            net = tf.slice(net, [0,0],[bs, -1])
+            print('net is', net)
+            return net
+
+def gradient_penalty(gan, gradient_penalty):
+    x = gan.graph.x
```
Comment (martyn) [inline hypergan/losses/common.py]: This doesn't work with resize_conv.  deconv works though.

**Comment by martyn** [comment]
Going to merge this and leave the checklist for pre-release.
